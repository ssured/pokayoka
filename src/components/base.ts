// from https://github.com/rebassjs/rebass/blob/89fc33f87e062e716092dc7bfbfcd4f300c014af/src/index.js
// moved from styled components to emotion

// import React from 'react';
import styled from '@emotion/styled';
import {
  space,
  color,
  width,
  height,
  flex,
  order,
  alignSelf,
  fontSize,
  fontFamily,
  fontWeight,
  textAlign,
  lineHeight,
  letterSpacing,
  borders,
  borderColor,
  borderRadius,
  buttonStyle,
  boxShadow,
  backgroundImage,
  backgroundSize,
  backgroundPosition,
  backgroundRepeat,
  opacity,
  variant,
} from 'styled-system';

import { Box, Flex } from '@rebass/grid/emotion';

export { Box, Flex };
// const css = props => props.css
// const themed = key => props => props.theme[key]

export const Text = styled(Box)(
  fontFamily,
  fontWeight,
  textAlign,
  lineHeight,
  letterSpacing
  //   themed('Text')
);

export const Heading = styled(Text)();
//   themed('Heading')

Heading.defaultProps = ({
  as: 'h2',
  m: 0,
  fontSize: 4,
  fontWeight: 'bold',
} as unknown) as any;

export const Link = styled(Box)();
//   themed('Link')

Link.defaultProps = ({
  as: 'a',
  color: 'blue',
} as unknown) as any;

export const Button = styled(Box)(
  {
    appearance: 'none',
    display: 'inline-block',
    textAlign: 'center',
    lineHeight: 'inherit',
    textDecoration: 'none',
  },
  fontWeight,
  borders,
  borderColor,
  borderRadius,
  buttonStyle
  //   themed('Button')
);

Button.defaultProps = ({
  as: 'button',
  fontSize: 'inherit',
  fontWeight: 'bold',
  m: 0,
  px: 3,
  py: 2,
  color: 'white',
  bg: 'blue',
  border: 0,
  borderRadius: 4,
} as unknown) as any;

export const Image = styled.img(
  {
    boxSizing: 'border-box',
    maxWidth: '100%',
    height: 'auto',
  },
  space,
  width,
  fontSize,
  color,
  flex,
  order,
  alignSelf,
  //   themed('Box'),
  //   css
  height,
  borderRadius
  //   themed('Image')
);

Image.defaultProps = ({
  as: 'img',
  m: 0,
} as unknown) as any;

const cards = variant({ key: 'cards' });

export const Card = styled(Box)(
  borders,
  borderColor,
  borderRadius,
  boxShadow,
  backgroundImage,
  backgroundSize,
  backgroundPosition,
  backgroundRepeat,
  opacity,
  //   themed('Card'),
  cards
);
