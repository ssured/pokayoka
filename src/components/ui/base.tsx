import styled from '@emotion/styled';

import {
  space,
  SpaceProps,
  color,
  ColorProps,
  width,
  WidthProps,
  fontSize,
  FontSizeProps,
  flexWrap,
  FlexWrapProps,
  flexDirection,
  FlexDirectionProps,
  alignItems,
  AlignItemsProps,
  justifyContent,
  JustifyContentProps,
  flex,
  FlexProps,
  order,
  OrderProps,
  alignSelf,
  AlignSelfProps,
  boxShadow,
  BoxShadowProps,
} from 'styled-system';

export const Box = styled<
  'div',
  SpaceProps &
    ColorProps &
    WidthProps &
    FontSizeProps &
    FlexProps &
    OrderProps &
    AlignSelfProps &
    BoxShadowProps
>('div')(
  {
    boxSizing: 'border-box',
  },
  space,
  color,
  width,
  fontSize,
  flex,
  order,
  alignSelf,
  boxShadow
);
Box.displayName = 'Box';

export const Flex = styled<
  typeof Box,
  FlexWrapProps & FlexDirectionProps & AlignItemsProps & JustifyContentProps
>(Box)(
  {
    display: 'flex',
  },
  flexWrap,
  flexDirection,
  alignItems,
  justifyContent
);

Flex.displayName = 'Flex';
