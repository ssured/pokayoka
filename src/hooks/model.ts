import { useState, useEffect } from 'react';
import {
  IAnyModelType,
  Instance,
  SnapshotIn,
  destroy,
  SnapshotOut,
  getSnapshot,
} from 'mobx-state-tree';
import { useDoc } from '../contexts/pouchdb';
import { merge, HamValue } from '../mst-ham/merge';
import { HAM_PATH } from '../mst-ham/index';
import { winningRev } from '../utils/pouchdb';
import { notifyUpdate } from '../global';

const getMachineState = () => Date.now();

const mergePouchDoc = <
  T extends { [HAM_PATH]: HamValue } & PouchDB.Core.IdMeta &
    PouchDB.Core.RevisionIdMeta,
  U extends { [HAM_PATH]: HamValue } & PouchDB.Core.IdMeta &
    PouchDB.Core.RevisionIdMeta
>(
  incoming: T,
  current: U
) => {
  if (incoming._id !== current._id) {
    throw new Error('cannot merge different documents');
  }
  const { resultHam, resultValue } = merge(
    getMachineState(),
    incoming['#'],
    incoming,
    current['#'],
    current
  );
  return {
    ...resultValue,
    [HAM_PATH]: resultHam,
    _rev: winningRev(incoming._rev, current._rev),
  };
};

const upsert = <
  T extends { [HAM_PATH]: HamValue } & PouchDB.Core.IdMeta &
    PouchDB.Core.RevisionIdMeta
>(
  db: PouchDB.Database,
  doc: T
) =>
  db.put(doc).catch(err => {
    if (err.status !== 409) throw err;

    return db
      .get(doc._id)
      .then(incoming => db.put(mergePouchDoc(incoming as any, doc)));
  });

export const useModel: <T extends IAnyModelType>(
  Model: T,
  db: PouchDB.Database,
  id: string
) => Instance<T> | null = (Model, db, id) => {
  const doc = useDoc<SnapshotIn<typeof Model>>(db, id);
  const [instance, setInstance] = useState<Instance<typeof Model> | null>(
    /*cache.has(id) ? Model.create(cache.get(id)) : */ null
  );

  useEffect(() => {
    if (doc == null) return;
    if (instance == null) {
      setInstance(
        Model.create(doc, {
          db,
          onSnapshot(snapshot: SnapshotOut<typeof Model>) {
            console.log('put snapshot', snapshot);
            notifyUpdate(snapshot);
            // upsert(db, snapshot).then(result => {
            //   instance && instance._setRev(result.rev);
            // });
          },
        })
      );
    } else {
      // console.log('will merge doc', doc);
      instance.merge(doc);
      // console.log(JSON.stringify(getSnapshot(instance), null, 2));
      // merge with the current doc
      // applySnapshot(instance, doc);
    }
    //   return () => {};
  }, [doc]);

  useEffect(() => {
    return () => {
      if (instance) {
        // cache.set(getIdentifier(instance), getSnapshot(instance));
        destroy(instance);
      }
    };
  }, []);

  return instance;
};
