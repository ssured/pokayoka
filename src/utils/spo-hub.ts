import mlts from 'monotonic-lexicographic-timestamp';
import { state, subj, pred, Tuple } from './spo';

export const getLocalState = mlts() as () => state;

export interface GetMessage {
  subj: subj;
  pred?: pred;
}

export interface StampedGetMessage extends GetMessage {
  type: 'get';
  localState: state;
}

export interface PutMessage {
  tuple: Tuple;
}
export interface StampedPutMessage extends PutMessage {
  type: 'put';
  localState: state;
}

type OutMessage = StampedGetMessage | StampedPutMessage;

type Listener = (msg: OutMessage) => void;

function whenMainThreadAvailable(effect: () => void) {
  setTimeout(effect, 0);
}

export class SPOHub {
  protected listeners: Set<Listener> = new Set();
  protected sources: WeakMap<Listener, unknown> = new WeakMap();

  constructor(public getCurrentState: () => state = getLocalState) {}

  public register(source: unknown, listener: Listener) {
    this.listeners.add(listener);
    this.sources.set(listener, source);

    return () => {
      this.listeners.delete(listener);
    };
  }

  protected fire(outMsg: OutMessage, skipSource?: unknown) {
    whenMainThreadAvailable(() => {
      // console.log('Hub fires', JSON.stringify(outMsg));

      this.listeners.forEach(listener => {
        if (this.sources.get(listener) !== skipSource) {
          listener(outMsg);
        }
      });
    });
  }

  public put(msg: PutMessage, skipSource?: unknown) {
    this.fire(
      { ...msg, type: 'put', localState: getLocalState() } as StampedPutMessage,
      skipSource
    );
  }

  public get(msg: GetMessage, skipSource?: any) {
    this.fire(
      { ...msg, type: 'get', localState: getLocalState() } as StampedGetMessage,
      skipSource
    );
  }
}
