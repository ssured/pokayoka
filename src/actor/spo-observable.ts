import { Actor, lookup } from '../utils/Actor';
import { SPOMessageType, objt } from './types';
import dset from 'dset';
import { observable, runInAction } from 'mobx';
import { ensureNever } from '../utils/index';
import {
  MessageType as PubSubMessageType,
  PublishMessage,
} from './spo-pub-sub';

declare global {
  interface ActorMessageType {
    spoObservable: Message;
    // dbState: SPOMessage;
  }
}

export type Message = PublishMessage;

export interface State {
  [key: string /* path [...subj, pred] */]: State | objt;
}

export class SPOObservableActor extends Actor<Message> {
  private pubsubActor = lookup('spoPubSub');
  public state = observable<State>({});

  async init() {
    // Subscribe to state updates
    this.pubsubActor.send({
      actorName: this.actorName!,
      type: PubSubMessageType.SUBSCRIBE,
    });
  }

  async onMessage(msg: Message) {
    switch (msg.type) {
      case PubSubMessageType.PUBLISH:
        const { payload } = msg;
        switch (payload.type) {
          case SPOMessageType.REQUEST:
            // ignore requests
            break;
          case SPOMessageType.TUPLE:
            // write to the local observable state
            runInAction(() => {
              dset(this.state, payload.s.concat(payload.p), payload.o);
            });
            break;
          default:
            ensureNever(payload);
        }
        break;
      default:
        ensureNever(msg.type);
    }
  }

  public request(id: string) {
    this.pubsubActor.send({
      type: PubSubMessageType.PUBLISH,
      payload: {
        type: SPOMessageType.REQUEST,
        s: [id],
      },
      sourceActorName: this.actorName!,
    });
  }
}
