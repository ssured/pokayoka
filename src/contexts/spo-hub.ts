import { createContext, useContext } from 'react';
import ReconnectingWebSocket from 'reconnecting-websocket';
import { SPOHub } from '../utils/spo-hub';
import { createObservable } from '../utils/spo-observable';
import { SPOStorage } from '../utils/spo-storage';
import { SPOWs } from '../utils/spo-ws';
import { SpotDB } from '../utils/spotdb';
import { useAuthentication } from './authentication';
import { many } from '../model/base';
import { userRelations } from '../model/User';

const spotDb = new SpotDB('pokayoka');

const hub = new SPOHub();
const storage = new SPOStorage(hub, spotDb);
const ws = new ReconnectingWebSocket(`ws://localhost:3000/spows`);
const server = new SPOWs(hub, ws);

const spo = createObservable<{
  [key: string]: User;
}>(hub, many(userRelations));

const SPOContext = createContext(spo);

export const useRoot = () => {
  const auth = useAuthentication();
  return useContext(SPOContext)[
    (auth.authentication &&
      auth.authentication.ok &&
      auth.authentication.name) ||
      'anonymous'
  ];
};

// export const useQuery = (query: Parameters<typeof spotDb.query>[0]) => {
//   return useMemo(() => {
//     const resultArray = observable.array<
//       ReturnType<typeof spotDb.query> extends AsyncIterableIterator<infer U>
//         ? U
//         : never
//     >([], { deep: false });
//     (async () => {
//       for await (const result of spotDb.query(query)) {
//         runInAction(() => resultArray.push(result));
//       }
//     })();
//     return resultArray;
//   }, []);
// };

// // export const useSubject =

// export const useModel = <T extends SPOShape, U>(
//   asyncFactory: (obj: SPOShape) => WrapAsync<T, U>,
//   id: string | subj
// ) => {
//   const shape = useContext(SPOContext);
//   const auth = useAuthentication();
//   const subj = [
//     'user',
//     (auth.authentication.ok && auth.authentication.name) || 'anonymous',
//   ].concat(id);
//   return useMemo(() => asyncFactory(shape.get(subj)), [
//     asyncFactory,
//     shape,
//     ...subj,
//   ]);
// };

// export const useAccount = () => useModel(AsyncUser, []);
