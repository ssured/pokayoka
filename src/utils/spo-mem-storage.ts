import { SPOHub, StampedGetMessage, StampedPutMessage } from './spo-hub';
import { ham } from './ham';
import dlv from 'dlv';
import { objt, state, subj } from './spo';
import dset from 'dset';

export class SPOMemStorage {
  private disposer: () => void;
  public data: any = {};

  constructor(protected hub: SPOHub) {
    this.disposer = hub.register(this, msg => {
      // @ts-ignore
      this[msg.type](msg);
    });
  }

  protected async get(msg: StampedGetMessage) {
    const { subj, pred } = msg;
    if (pred) {
      const [objt, state]: [objt, state] =
        dlv(this.data, subj.concat(pred)) || ([] as any);
      if (objt) {
        this.hub.put({ tuple: [subj, pred, objt, state] }, this);
      }
      return;
    }

    const walk = (subj: subj) => {
      const value = dlv(this.data, subj) || {};

      if (value && typeof value === 'object') {
        Object.entries(value).forEach(([pred, val]) => {
          if (Array.isArray(val)) {
            const [objt, state] = val as [objt, state];
            this.hub.put({ tuple: [subj, pred, objt, state] }, this);
          } else if (val && typeof val === 'object') {
            walk(subj.concat(pred));
          }
        });
      }
    };

    walk(subj);
  }

  protected async put(msg: StampedPutMessage) {
    const [subj, pred, incomingValue, incomingState] = msg.tuple;
    const [currentValue = undefined, currentState = ''] =
      dlv(this.data, subj.concat(pred)) || ([] as any);
    const machineState = msg.localState;

    const result = ham(
      machineState,
      incomingState,
      currentState,
      incomingValue,
      currentValue
    );

    if (result.resolution === 'merge' && result.incoming) {
      dset(this.data, subj.concat(pred), [incomingValue, incomingState]);
    }
  }

  public destroy() {
    this.disposer();
  }
}
