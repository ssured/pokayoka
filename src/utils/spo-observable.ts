import charwise from 'charwise';
import dlv from 'dlv';
import dset from 'dset';
import { RelationsOf } from '../model/base';
import { createConvergeFunction } from './ham';
import { ensureNever } from './index';
import { isObjt, objt, pred, SPOShape, subj } from './spo';
import { SPOHub } from './spo-hub';
import { createUniverse, getPath, Maybe, ThunkTo } from './universe';

const pathToKey = charwise.encode;
// @ts-ignore
window.charwise = charwise;

type RequestedMap = { [key: string]: boolean | RequestedMap };

export function createObservable<T extends SPOShape = SPOShape>(
  hub: SPOHub,
  runtimeShape: RelationsOf<T>
): ThunkTo<T> {
  type state = string;
  type SubjStateMap = Record<string, state>;

  const requested: RequestedMap = {};

  function request(subj: subj) {
    // if a more generic subject was already requested, ignore the request
    try {
      dset(requested, subj, true);
      // console.log('spo-observable get', subj);
      hub.get({ subj }, root);
    } catch (e) {
      if (e.message.match(/^cannot create property/i) == null) {
        throw e;
      }
    }
  }

  const subjStates: Record<string, SubjStateMap> = {};

  const getStateMap = (subj: subj): SubjStateMap =>
    subjStates[pathToKey(subj)] || (subjStates[pathToKey(subj)] = {});

  // merge data into this observable, respecting HAM
  const mergeTuple = ([Si, Di]: [state, objt], [subj, pred]: [subj, pred]) => {
    const states = getStateMap(subj);

    const Sc = states[pred] || '';

    // FIXME: expose current value as helper function of universe library
    const Dc = ((dlv<Maybe<SPOShape>>(root(), subj) || {})[pred] ||
      null) as objt;

    const result = converge([Sc, Dc], [Si, Di], [subj, pred]);

    // console.log('spo-observable merge', subj, pred, result);
  };

  const converge = createConvergeFunction<state, objt, [subj, pred]>(
    hub.getCurrentState,
    {
      saveFuture: (incoming, meta) => {
        const waitMs = 5000; // FIXME implement calculation msFromNow(incomingState)
        setTimeout(() => mergeTuple(incoming, meta), waitMs);
      },
      saveHistorical: () => {},
      saveNow: ([state, data], [subj, pred]) => {
        const states = getStateMap(subj);
        states[pred] = state;
        // console.log(
        //   `saveNow ${subj.join('/')} ${pred} = ${JSON.stringify(data)}`
        // );
        set(subj, false, { [pred]: data });
      },
    }
  );

  // create the main root
  const { root, set /*, get*/ } = createUniverse<T>({
    runtimeShape,
    resolve: subj => {
      subjStates[pathToKey(subj)] = {
        ...(subjStates[pathToKey(subj)] || { states: {} }),
      };

      return {
        onActive: () => {
          request(subj);
        },
      };
    },
    updateListener: function updateListener(subj, value) {
      const states = getStateMap(subj);

      for (const [pred, objtOrReferencedValue] of Object.entries(value)) {
        // check if the object is a reference, if so, use its path as objt
        const objt = getPath(objtOrReferencedValue) || objtOrReferencedValue;

        if (isObjt(objt)) {
          const state = hub.getCurrentState();
          states[pred] = state;
          hub.put(
            { tuple: [subj, pred, objt == null ? null : objt, state] },
            root
          );
        } else {
          updateListener(subj.concat(pred), objt);
        }
      }
    },
    pathToKey,
  });

  hub.register(root, async msg => {
    switch (msg.type) {
      case 'put': {
        const {
          tuple: [subj, pred, Di, Si],
        } = msg;
        mergeTuple([Si, Di], [subj, pred]);

        break;
      }
      case 'get': {
        // ignored as runtime is not a data store
        break;
      }
      default:
        ensureNever(msg);
    }
  });

  return root;
}
