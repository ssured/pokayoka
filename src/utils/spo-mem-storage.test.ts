import { many, RelationsOf } from '../model/base';
import { SPOHub } from './spo-hub';
import { SPOMemStorage } from './spo-mem-storage';
import { createObservable } from './spo-observable';

const delay = (ms: number) => new Promise(res => setTimeout(res, ms));

describe('mem based SPO observable test', () => {
  let hub: SPOHub;
  let storage: SPOMemStorage;
  let state: number;
  let getCurrentState: () => string;

  beforeEach(() => {
    state = 0;
    getCurrentState = (() => {
      let currentState = state;
      let subState = 0;
      return () => {
        if (state > currentState) {
          currentState = state;
          subState = 0;
        }
        return [
          `000${currentState}`.substr(-3),
          `000${(subState += 1)}`.substr(-3),
        ].join('.');
      };
    })();
    hub = new SPOHub(getCurrentState);
    storage = new SPOMemStorage(hub);
  });

  test('local state generator', () => {
    expect(getCurrentState()).toBe('000.001');
    expect(getCurrentState()).toBe('000.002');
    state += 1;
    expect(getCurrentState()).toBe('001.001');
    expect(getCurrentState()).toBe('001.002');
  });

  test.only('basic', async () => {
    type User = { value: string };
    const userRelations: RelationsOf<User> = {};

    type Root = { [key: string]: User };
    const spo = createObservable<Root>(hub, many(userRelations));

    spo().test = { value: 'test' };
    await delay(1); // run async stuff

    expect(storage.data).toMatchSnapshot();

    // build a reference
    spo().testRef = spo().test;
    await delay(2); // run async stuff

    expect(storage.data).toMatchSnapshot();
    // a reference is stored instead of the data
    expect(Array.isArray(storage.data.testRef[0])).toBeTruthy();
    expect(spo().testRef.value).toBe('test');

    // update the source and check the reference
    spo().test.value = 'update';
    await delay(1); // run async stuff

    expect(spo().testRef.value).toBe('update');
  });

  test('nested', async () => {
    type Person = {
      identifier: string;
      name: string;
    };
    const personRelations: RelationsOf<Person> = {};
    type User = {
      identifier: string;
      name: string;
      is: Person;
    };
    const userRelations: RelationsOf<User> = {
      is: personRelations,
    };

    type Root = {
      [key: string]: User;
    };

    const spo = createObservable<Root>(hub, many(userRelations));

    spo().test1 = {
      identifier: 'test1',
      name: 'one',
      is: {
        identifier: 'p1',
        name: 'test1name',
      },
    };

    state += 1; // ?

    spo().test2 = {
      identifier: 'test2',
      name: 'two',
      is: spo.test1.is(),
    };

    await delay(1);

    expect(storage.data).toMatchSnapshot();
    expect(spo.test2.is().name).toBe('test1name');
  });
});
