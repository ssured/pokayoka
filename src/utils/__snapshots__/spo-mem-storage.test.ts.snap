// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mem based SPO observable test basic 1`] = `
Object {
  "test": Object {
    "value": Array [
      "test",
      "000.001",
    ],
  },
}
`;

exports[`mem based SPO observable test basic 2`] = `
Object {
  "test": Object {
    "value": Array [
      "test",
      "000.001",
    ],
  },
  "testRef": Array [
    Array [
      "test",
    ],
    "000.002",
  ],
}
`;

exports[`mem based SPO observable test nested 1`] = `
Object {
  "test1": Object {
    "identifier": Array [
      "test1",
      "000.001",
    ],
    "is": Object {
      "identifier": Array [
        "p1",
        "000.005",
      ],
      "name": Array [
        "test1name",
        "000.006",
      ],
    },
    "name": Array [
      "one",
      "000.002",
    ],
  },
  "test2": Object {
    "identifier": Array [
      "test2",
      "001.001",
    ],
    "is": Array [
      Array [
        "test1",
        "is",
      ],
      "001.003",
    ],
    "name": Array [
      "two",
      "001.002",
    ],
  },
}
`;
